File: Java/Portscan.java

import java.net.Socket;
import java.net.InetSocketAddress;
import java.io.IOException;


/*
   Run :
     - Method :
        - javac Portscan.java
        - java Portscan (name of the class )
        - for this code :
           - java Portscan <host>(www.google.com or 192.168.224.1)
           - time java Portscan <host>(www.gmail.com or 4.4.4.4)
     - Other method :
       - alias run='javac Portscan.java; java Portscan';
       - run
       - for this code :
          - run <host>(www.youtube.com or 8.8.8.8)
          - time run <host>(www.facebook.com or 157.548.123.2)
*/
class Portscan {

    private static void printUsage(){
      System.out.println("Usages : ");
      System.out.println("  Java Portscan <host>>");
      System.out.println("Examples : ");
      System.out.println("  Java Portscan 127.0.0.1");
      System.out.println("  Java Portscan www.google.com");

    }

    private static void checkPort(String host, Integer portNumber){
      //open a socket,
     // connect to host/post
     //()->
     Integer timeout = 3000;
     Thread thread = new Thread(()->{
       Socket clientSocket = new Socket();
       try {
          clientSocket.connect(new InetSocketAddress(host, portNumber), timeout);
          System.out.printf("[-] %d connected %s ",portNumber,host);
          clientSocket.close();
       }catch (IOException e) {

       }
     });
     thread.start();

    }

    // public static void main function
    public static void main(String [] args){
        // args.length == 0 display printUsage() info
        if (args.length == 0){
          System.out.println("No arguments received.");
          printUsage();
          System.exit(1);
        }
        // else display this info
        String host = args[0];
        System.out.println(args);
        System.out.println(host);
        System.out.printf("Scanning %s now ... \n", host);

        //port 1-65535
        // open TCP connection
        for(Integer portNumber=0; portNumber<=65535; portNumber++){
           checkPort(host, portNumber);
        }
    }
}

// alias run='javac Portscan.java; java Portscan'
// echo $?
// 41:11

File: Go/portScan.go
package main

import (
	//"fmt"
	"log"
	"net"
	"os"
	"strconv"
	"time"
)

func printUsage() {
	log.Println("Usage : ")
	log.Println("  go run fileName.go <host>")
	log.Println("Example : ")
	log.Println("  go run portScan.go www.google.com")
	log.Println("  go run portScan.go 8.8.8.8")
}

func testTcpConnect(host string, port int, doneChannel chan bool) {

	timeoutLength := 5 * time.Second
	conn, err := net.DialTimeout("tcp ", host+": "+strconv.Itoa(port), timeoutLength)

	if err != nil {
		doneChannel <- false
		return
	}
	conn.Close()
	log.Printf("[+] %d connected ", port)
	doneChannel <- true
}

func main() {

	if len(os.Args) == 1 {
		log.Println("No arguments received.")
		printUsage()
		os.Exit(1)
	}
	// Take a host from args1
	// if no args provided, print usage
	// host = args1

	// for ports 1-65535
	//   go testTcpConnect()

	doneChannel := make(chan bool)
	activeThreadCount := 0
	log.Println("Scanning host : " + os.Args[1])
	for portNumber := 1; portNumber <= 65535; portNumber++ {
		activeThreadCount++
		go testTcpConnect(os.Args[1], portNumber, doneChannel)
	}

	for {
		<-doneChannel
		activeThreadCount--
		// log.Printf("Reducing threadAcount %d ", activeThreadCount)
		if activeThreadCount == 0 {
			break
		}
	}
	log.Println("Done.")
	// until activeThreadCount == 0 keep checking
	//

}
